File: src/2025-02-23-deepclaude-try/go/exercises/practice/say/say.go
package say

import "strings"

var (
	underTwenty = []string{
		"",         // 0
		"one",      // 1
		"two",      // 2
		"three",    // 3
		"four",     // 4
		"five",     // 5
		"six",      // 6
		"seven",    // 7
		"eight",    // 8
		"nine",     // 9
		"ten",      // 10
		"eleven",   // 11
		"twelve",   // 12
		"thirteen", // 13
		"fourteen", // 14
		"fifteen",  // 15
		"sixteen",  // 16
		"seventeen",// 17
		"eighteen", // 18
		"nineteen", // 19
	}
	tens = []string{
		"",        // 0
		"ten",     // 10
		"twenty",  // 20
		"thirty",  // 30
		"forty",   // 40
		"fifty",   // 50
		"sixty",   // 60
		"seventy", // 70
		"eighty",  // 80
		"ninety",  // 90
	}
)

func Say(n int64) (string, bool) {
	if n < 0 || n > 999999999999 {
		return "", false
	}
	if n == 0 {
		return "zero", true
	}

	chunks := splitNumber(n)
	scales := []string{"", "thousand", "million", "billion"}
	var parts []string

	for i, chunk := range chunks {
		scaleIndex := len(chunks) - i - 1
		part := convertTrio(chunk)
		if part != "" {
			if scaleIndex > 0 {
				part += " " + scales[scaleIndex]
			}
			parts = append(parts, part)
		}
	}

	return strings.Join(parts, " "), true
}

func splitNumber(n int64) []int {
	var chunks []int
	for n > 0 {
		chunks = append(chunks, int(n%1000))
		n /= 1000
	}
	// Reverse chunks to get highest order first
	for i, j := 0, len(chunks)-1; i < j; i, j = i+1, j-1 {
		chunks[i], chunks[j] = chunks[j], chunks[i]
	}
	return chunks
}

func convertTrio(n int) string {
	if n == 0 {
		return ""
	}
	var result string
	hundreds := n / 100
	if hundreds > 0 {
		result += underTwenty[hundreds] + " hundred"
	}
	tensPart := n % 100
	if tensPart > 0 {
		if hundreds > 0 {
			result += " "
		}
		if tensPart < 20 {
			result += underTwenty[tensPart]
		} else {
			tensDigit := tensPart / 10
			onesDigit := tensPart % 10
			result += tens[tensDigit]
			if onesDigit > 0 {
				result += "-" + underTwenty[onesDigit]
			}
		}
	}
	return result
}

File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/say/say.js
export const say = (n) => {
  const ones = ['', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];
  const teens = ['ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'];
  const tens = ['', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'];
  const scaleNames = ['', 'thousand', 'million', 'billion', 'trillion'];

  function convertTens(num) {
    if (num === 0) return '';
    if (num < 10) return ones[num];
    if (num < 20) return teens[num - 10];
    const ten = Math.floor(num / 10);
    const one = num % 10;
    return tens[ten] + (one !== 0 ? '-' + ones[one] : '');
  }

  function convertChunk(num) {
    if (num === 0) return '';
    const hundred = Math.floor(num / 100);
    const remainder = num % 100;
    let result = '';
    if (hundred > 0) {
      result += ones[hundred] + ' hundred ';
    }
    const tensStr = convertTens(remainder);
    if (tensStr) {
      result += tensStr;
    }
    return result.trim().replace(/\s+/g, ' ');
  }

  if (typeof n !== 'number' || n < 0 || n > 999999999999 || !Number.isInteger(n)) {
    throw new Error('Number must be between 0 and 999,999,999,999.');
  }

  if (n === 0) {
    return 'zero';
  }

  const chunks = [];
  let current = n;
  while (current > 0) {
    chunks.push(current % 1000);
    current = Math.floor(current / 1000);
  }
  const reversedChunks = chunks.reverse();

  const parts = [];
  for (let i = 0; i < reversedChunks.length; i++) {
    const chunk = reversedChunks[i];
    const chunkStr = convertChunk(chunk);
    if (chunkStr) {
      const scaleIndex = reversedChunks.length - 1 - i;
      const scale = scaleNames[scaleIndex];
      parts.push(chunkStr + (scale ? ` ${scale}` : ''));
    }
  }

  return parts.join(' ');
};
